In [70]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
import warnings
warnings.filterwarnings("ignore")

In [71]:
df = pd.read_csv('https://raw.githubusercontent.com/cmpe255-orion/car-price-prediction/main/listing.csv')
del df['Unnamed: 0']
print(df.shape)
df.head(5)

(9936, 20)


year       make     model  ... owners      use_type   cpo
0  2021  Chevrolet  Suburban  ...      1  Personal Use  True
1  2017   Cadillac       CTS  ...      1  Personal Use  True
2  2019        Kia     Forte  ...      2  Personal Use  True
3  2018       Ford     F-150  ...      1  Personal Use  True
4  2018       Ford     F-150  ...      4    Rental Use  True

[5 rows x 20 columns]

This dataset scraped from TrueCar has 9936 car listings with 20 columns. The following are the column descriptions:
<pre>
1. year:            Model Year
2. make:            Manufacturer
3. model:           Vehicle Model
4. sub_model:       Vehicle Model variant
5. city:            City where the car is located
6. state:           State where the car is located
7. mileage:         Car's current odometer reading
8. price:           Car's selling price
9. style:           Style of the car
10. exterior_color: Car's exterior color
11. interior_color: Car's interior color
12. engine:         Car's engine specification
13. drive_type:     Drive Type
14. fuel_type:      Type of fuel
15. transmission:   Type of transmission
16. accidents:      Number of accidents
17. title:          Type of title(clean or salvaged, etc..)
18. owners:         Number of previous owners
19. use_type:       Type of usage of the car
20. cpo:            Certified Pre-Owned or Not
</pre>



**Check for unique values**

In [72]:
print(df['style'].unique())
print(df['drive_type'].unique())
print(df['fuel_type'].unique())
print(df['transmission'].unique())
print(df['accidents'].unique())
print(df['title'].unique())
print(df['owners'].unique())
print(df['use_type'].unique())
print(df['cpo'].unique())

['SUV' 'Sedan' 'Pickup Truck' 'Minivan' 'Coupe' 'Hatchback' 'Cargo Van'
 'Convertible' 'Passenger Van' 'Wagon' 'Chassis Cab Truck']
['RWD' 'AWD' 'FWD' '4WD']
['Gas' 'Hybrid' 'Diesel' 'Electric' 'Plug-In Hybrid' 'Hydrogen']
['Automatic' 'Manual']
[0 1 2 3 5 4]
['Clean' 'Salvage History' 'Theft History' 'Lemon History']
[ 1  2  4  3  0  5  7  6  8  9 10]
['Personal Use' 'Rental Use' 'Mixed Use' 'Fleet Use']
[ True]


Here we find that cpo has only one value 'True'. Probably that field has wrong data. Hence it can be removed.

In [73]:
del df['cpo']
df.shape

(9936, 19)

**Check for missing values**

In [74]:
df.isna().sum()

year              0
make              0
model             0
sub_model         0
city              0
state             0
mileage           0
price             0
style             0
exterior_color    0
interior_color    0
engine            0
drive_type        0
fuel_type         0
transmission      0
accidents         0
title             0
owners            0
use_type          0
dtype: int64

We find that there are no missing values

**Check the data type of the columns**

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9936 entries, 0 to 9935
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   year            9936 non-null   int64 
 1   make            9936 non-null   object
 2   model           9936 non-null   object
 3   sub_model       9936 non-null   object
 4   city            9936 non-null   object
 5   state           9936 non-null   object
 6   mileage         9936 non-null   object
 7   price           9936 non-null   object
 8   style           9936 non-null   object
 9   exterior_color  9936 non-null   object
 10  interior_color  9936 non-null   object
 11  engine          9936 non-null   object
 12  drive_type      9936 non-null   object
 13  fuel_type       9936 non-null   object
 14  transmission    9936 non-null   object
 15  accidents       9936 non-null   int64 
 16  title           9936 non-null   object
 17  owners          9936 non-null   int64 
 18  use_type

We could conver the fields mileage to numeric and remove the $ sign and comma from the price.

In [76]:
df['mileage'] = df['mileage'].apply(lambda x: ''.join(x.split(',')))
df['mileage'] = df['mileage'].astype(int)
df['mileage'].head()

0      8404
1     29342
2     17593
3    195205
4     47922
Name: mileage, dtype: int64

In [77]:
df['price'] = df['price'].str.replace(r'$', '')
df['price'] = df['price'].str.replace(r',', '')
df['price'].head()

0    67994
1    37000
2    22395
3    20995
4    32988
Name: price, dtype: object

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9936 entries, 0 to 9935
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   year            9936 non-null   int64 
 1   make            9936 non-null   object
 2   model           9936 non-null   object
 3   sub_model       9936 non-null   object
 4   city            9936 non-null   object
 5   state           9936 non-null   object
 6   mileage         9936 non-null   int64 
 7   price           9936 non-null   object
 8   style           9936 non-null   object
 9   exterior_color  9936 non-null   object
 10  interior_color  9936 non-null   object
 11  engine          9936 non-null   object
 12  drive_type      9936 non-null   object
 13  fuel_type       9936 non-null   object
 14  transmission    9936 non-null   object
 15  accidents       9936 non-null   int64 
 16  title           9936 non-null   object
 17  owners          9936 non-null   int64 
 18  use_type

**Check for duplicates and remove them**

In [79]:
print(df.duplicated().sum())
df.shape
df = df.drop_duplicates(keep='first')
df.shape

29


(9907, 19)

Apparently there were 29 duplicate records and we have removed them.

**Outlier Analysis**

Check the mileage and price for any discrepancy in values

In [85]:
display(df.sort_values(by='mileage',ascending=True).head(3))
display(df.sort_values(by='mileage',ascending=False).head(3))

year           make     model  ...  title owners      use_type
3072  2020         Toyota   Corolla  ...  Clean      1  Personal Use
3572  2019            BMW  7 Series  ...  Clean      1  Personal Use
404   2021  Mercedes-Benz   C-Class  ...  Clean      1  Personal Use

[3 rows x 19 columns]

year  make             model  ...  title owners      use_type
46    2014  Ford             F-150  ...  Clean      2  Personal Use
9453  2014   Ram  3500 Chassis Cab  ...  Clean      1  Personal Use
3629  2018   Ram              3500  ...  Clean      1  Personal Use

[3 rows x 19 columns]

By sorting the mileage data in both the ascending and descending order and looking for the minimum and maximum values, it does not look like there is any discrepancy in the data. A new car would have low miles and the lowest value which is 9 miles is nominal. In the same fashion the maximum value of 380309 miles for an old car is also very reasonable.

In [93]:
display(df.sort_values(by='price',ascending=False).head(5))

year     make    model  ...  title owners      use_type
6919  2018   Toyota    Camry  ...  Clean      1  Personal Use
9881  2018    Buick   Encore  ...  Clean      1  Personal Use
3628  2020   Toyota  Corolla  ...  Clean      1  Personal Use
372   2017  Hyundai  Elantra  ...  Clean      3     Mixed Use
8058  2015     Ford   Fusion  ...  Clean      2     Fleet Use

[5 rows x 19 columns]

When sorting by price we find that there are some records which have the pricing information unavailable. Hence remove those records.

In [96]:
df = df[df['price'] != 'No Price Available']

Convert the price to numeric and sort again

In [98]:
df['price'] = df['price'].astype(int)

In [99]:
display(df.sort_values(by='price',ascending=False).head(5))
display(df.sort_values(by='price',ascending=True).head(5))

year           make              model  ...  title owners      use_type
2139  2011    Rolls-Royce              Ghost  ...  Clean      4  Personal Use
7191  2017           Land  Rover Range Rover  ...  Clean      1  Personal Use
5906  2012        Porsche                911  ...  Clean      3  Personal Use
3413  2012          Aston      Martin Virage  ...  Clean      4  Personal Use
3108  2022  Mercedes-Benz                GLS  ...  Clean      0  Personal Use

[5 rows x 19 columns]

year        make       model  ...  title owners      use_type
1179  1998     Pontiac  Bonneville  ...  Clean      4    Rental Use
8308  2001       Buick       Regal  ...  Clean      5     Mixed Use
624   2001     Hyundai      Accent  ...  Clean      5  Personal Use
4709  2002   Chevrolet    Suburban  ...  Clean      3  Personal Use
7580  1998  Oldsmobile     Cutlass  ...  Clean      4  Personal Use

[5 rows x 19 columns]

The lowest price of 2000 and the highest price of 114995 is well within the acceptable range for car prices.

In [100]:
df.shape

(9904, 19)

In [101]:
df.to_csv('listing-cleaned.csv', encoding = 'utf-8')